### PR TITLE
add converter for DIRAC to convert4qmc

### PR DESCRIPTION
**Pull request by camelto2**
## Proposed changes

This PR adds an initial implementation for a converter from DIRAC to QMCPACK for molecular spin-orbit calculations. This adds a `-dirac` option to the convert4qmc executable which reads the *.out file from DIRAC and generates qmcpack xml files as well as the hdf5 file required to store the spinors. 

For the time being, this converter only stores the spinors and generates a single slater determinant wave function, whereas most DIRAC calculations involve CI wave function in order to get the correct spin-orbit split states. I will add this functionality in a future PR, in the mean time the xml wavefunction will have to be modified by hand to get the correct wavefunction. This will currently work for single determinant closed shell wave functions however. 

## What type(s) of changes does this code introduce?

- New feature
- Documentation changes

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
my workstation, gnu 9.2.0
## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes. Documentation has been added (if appropriate)

### CODE REVIEW

File: tests/converter/test_Bi_dirac/gold.wfnoj.xml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0"?>
+<qmcsystem>
+  <wavefunction name="psi0" target="e">
+    <determinantset type="MolecularOrbital" name="LCAOBSet" source="ion0" transform="yes" href="test.orbs.h5">
+      <slaterdeterminant>
+        <determinant id="updet" size="5">
+          <occupation mode="ground"/>
+          <coefficient size="216" spindataset="0"/>
+        </determinant>
+        <basisset/>
```
Comment (ye-luo) [inline tests/converter/test_Bi_dirac/gold.wfnoj.xml:10]: I saw a dangling `<basisset/>` not sure if the issue already exists in upstream.

**Review by ye-luo — COMMENTED** [review]
Do you mind we merge with squash? A lot of commits are merge commits.  Tries to concentrate. Feel free to squash or rebase if you would like to do it yourself.

File: tests/converter/test_Bi_dirac/gold.wfnoj.xml (hunk 1 above)
Comment (camelto2) [reply tests/converter/test_Bi_dirac/gold.wfnoj.xml:10]: looking into this now. this shouldn't be there since it is stored in the hdf5.

Comment (camelto2) [reply tests/converter/test_Bi_dirac/gold.wfnoj.xml]: fixed

**Comment by ye-luo** [comment]
What is the pass rate of `ctest -L converter -j8 -E qe`?

**Comment by camelto2** [comment]
> ctest -L converter -j8 -E qe

On my workstation, all pass

**Comment by ye-luo** [comment]
> > What is the pass rate of ctest -L converter -j8 -E qe?
> 
> On my workstation, all pass

Cool. In addition, there is an uninitialized value warning and ubsan CI is failing.

**Comment by camelto2** [comment]
> Do you mind we merge with squash? A lot of commits are merge commits. Tries to concentrate. Feel free to squash or rebase if you would like to do it yourself.

I don't mind. However, I will profess my ignorance here and state that I'm not sure how to do this correctly. My current attempts on my local branch keep resulting in failures so I'm likely doing something wrong

**Comment by ye-luo** [comment]
If a simple rebase on develop doesn't work. We just do squash at the final merge. No need of further struggle with it. To keep a branch clean, frequent rebase is needed. Fixing it at a late stage is generally hard.

**Comment by camelto2** [comment]
> Cool. In addition, there is an uninitialized value warning and ubsan CI is failing.

the ubsan isn't failing based on the recent commit. my local build isn't giving any warning about uninitialized values...which value isn't being initialized?

**Comment by ye-luo** [comment]
I saw the warning in 5ec5ccc https://github.com/QMCPACK/qmcpack/runs/3194741405#step:5:581
you last commit was not showing the same issue but the change doesn't seem like to fix the warning. Strange.

**Comment by camelto2** [comment]
> I saw the warning in [5ec5ccc](https://github.com/QMCPACK/qmcpack/commit/5ec5ccce984710cbb06889065c8e58b649099743) https://github.com/QMCPACK/qmcpack/runs/3194741405#step:5:581
> you last commit was not showing the same issue but the change doesn't seem like to fix the warning. Strange.
```

/__w/qmcpack/qmcpack/src/QMCTools/QMCGaussianParserBase.cpp:641:21: warning: variable 'ddet' is used uninitialized whenever 'for' loop exits because its condition is false [-Wsometimes-uninitialized]
    for (int i = 0; i < numMO; i++)
                    ^~~~~~~~~
/__w/qmcpack/qmcpack/src/QMCTools/QMCGaussianParserBase.cpp:687:28: note: uninitialized use occurs here
  cur = xmlAddSibling(cur, ddet);
                           ^~~~
/__w/qmcpack/qmcpack/src/QMCTools/QMCGaussianParserBase.cpp:641:21: note: remove the condition if it is always true
    for (int i = 0; i < numMO; i++)
                    ^~~~~~~~~
/__w/qmcpack/qmcpack/src/QMCTools/QMCGaussianParserBase.cpp:617:18: note: initialize the variable 'ddet' to silence this warning
  xmlNodePtr ddet;
                 ^
                  = nullptr
1 warning generated.
```

the uninitialized use seemed due to the `  cur = xmlAddSibling(cur, ddet);` line. the fix for the basisset printed had to do with this. Since we don't need a downdet for the spinors, I made an conditional on printing the downdet based on whether or not we have spinors, but I left this line outside. That ultimately resulted in the printing of the `<basisset/>` and therefore ddet was uninitialized. 

Fixing the basisset printing should have fixed that...you are still seeing the warning?

**Comment by ye-luo** [comment]
@camelto2 then everything is good. CI is accurate.

**Review by ye-luo — APPROVED** [review]

**Comment by ye-luo** [comment]
Test this please

**Comment by camelto2** [comment]
The failing tests all involve the cuda code, which this code shouldn't have anything to do with. The failures are all due to the 
`h5diff: <test.orbs.h5>: unable to open file`

I am also unable to reproduce this failure on my workstation. I rebuilt the new converter using the same CMake arguments as some of the failing tests (-DQMC_CUDA=1) and the convert4qmc generates the test.orbs.h5 and converter_test.py successfully finishes. So I'm not sure how to track down the error

**Comment by williamfgc** [comment]
@camelto2 I built your branch on sulfur and I can reproduce. It looks like `test.orbs.h5` doesn't get generated under `tests/converter/test_Bi_dirac` , other converter tests generated it in other directories. Hope the logs below help debug.

```
$ ctest -R converter
Test project /home/wfg/qmcpack/build
      Start 40: converter_test_He_sto3g
 1/18 Test #40: converter_test_He_sto3g ..........   Passed    0.06 sec
      Start 41: converter_test_Be_sto3g
 2/18 Test #41: converter_test_Be_sto3g ..........   Passed    0.33 sec
      Start 42: converter_test_Be_ccd
 3/18 Test #42: converter_test_Be_ccd ............   Passed    0.06 sec
      Start 43: converter_test_O_ext
 4/18 Test #43: converter_test_O_ext .............   Passed    0.14 sec
      Start 44: converter_test_C_sto3g
 5/18 Test #44: converter_test_C_sto3g ...........   Passed    0.35 sec
      Start 45: converter_test_HCNp
 6/18 Test #45: converter_test_HCNp ..............   Passed    0.06 sec
      Start 46: converter_test_aldet1
 7/18 Test #46: converter_test_aldet1 ............   Passed    0.05 sec
      Start 47: converter_test_aldet5
 8/18 Test #47: converter_test_aldet5 ............   Passed    0.05 sec
      Start 48: converter_test_aldet6
 9/18 Test #48: converter_test_aldet6 ............   Passed    0.07 sec
      Start 49: converter_test_HDF5_FeCO6
10/18 Test #49: converter_test_HDF5_FeCO6 ........   Passed    0.29 sec
      Start 50: converter_test_HDF5_Be_ccd
11/18 Test #50: converter_test_HDF5_Be_ccd .......   Passed    0.08 sec
      Start 51: converter_test_Bi_dirac
12/18 Test #51: converter_test_Bi_dirac ..........***Failed    0.38 sec
      Start 52: converter_test_NaCl_qbox
13/18 Test #52: converter_test_NaCl_qbox .........   Passed    0.11 sec
      Start 53: converter_test_O2_pwscf
14/18 Test #53: converter_test_O2_pwscf ..........   Passed    0.10 sec
      Start 54: converter_test_LiH_pyscf
15/18 Test #54: converter_test_LiH_pyscf .........   Passed    0.06 sec
      Start 55: converter_test_LiH_pyscf_UHF
16/18 Test #55: converter_test_LiH_pyscf_UHF .....   Passed    0.06 sec
      Start 56: converter_test_LiH_qp
17/18 Test #56: converter_test_LiH_qp ............   Passed    0.06 sec
      Start 57: converter_test_C4_MSD_excited
18/18 Test #57: converter_test_C4_MSD_excited ....   Passed    0.06 sec

94% tests passed, 1 tests failed out of 18

Label Time Summary:
converter        =   2.37 sec*proc (18 tests)
deterministic    =   2.37 sec*proc (18 tests)

Total Test time (real) =   2.39 sec

The following tests FAILED:
	 51 - converter_test_Bi_dirac (Failed)
Errors while running CTest
```

```
51: Test command: /home/wfg/qmcpack/build/tests/converter/converter_test.py "/home/wfg/qmcpack/build/tests/converter/test_Bi_dirac" "--exe" "/home/wfg/qmcpack/build/bin/convert4qmc" "--h5diff" "/usr/bin/h5diff"
51: Test timeout computed to be: 120
51: h5diff: <test.orbs.h5>: unable to open file
51: Return code nonzero:  -6
51: Stderr not emptry 
51: Unknown basis function type. Aborting
51: 
51: h5diff reported a difference
51:   FAIL
1/1 Test #51: converter_test_Bi_dirac ..........***Failed    0.58 sec

0% tests passed, 1 tests failed out of 1
```

```
$ find . -name *.orbs.h5 
./tests/wavefunction/LiH.orbs.h5
./tests/wavefunction/Bi.orbs.h5
./tests/converter/test_HDF5_FeCO6/gold.orbs.h5
./tests/converter/test_HDF5_FeCO6/test.orbs.h5
./tests/converter/test_HDF5_Be_ccd/gold.orbs.h5
./tests/converter/test_HDF5_Be_ccd/test.orbs.h5
./tests/converter/test_Bi_dirac/gold.orbs.h5
```

**Comment by camelto2** [comment]
@williamfgc Are you able to just run convert4qmc -nojastrow -prefix test -dirac df_Bi.out in the test directory? that is what the converter_test.py should be executing. I'm curious to see what the output of just the convert4qmc executable is

**Comment by williamfgc** [comment]
@camelto2 let me know if the below message makes sense (looks like a core dump, I'm working on those BTW):

```
"/home/wfg/qmcpack/build/bin/convert4qmc" -nojastrow -prefix test -dirac df_Bi.out
Index of ion charge 0
Index of valence charge 1
Using test to name output files
Error when opening file: df_Bi.out
Aborted (core dumped)
[wfg@sulfur build]$ valgrind "/home/wfg/qmcpack/build/bin/convert4qmc" -nojastrow -prefix test -dirac df_Bi.out
==3169344== Memcheck, a memory error detector
==3169344== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==3169344== Using Valgrind-3.16.0 and LibVEX; rerun with -h for copyright info
==3169344== Command: /home/wfg/qmcpack/build/bin/convert4qmc -nojastrow -prefix test -dirac df_Bi.out
==3169344== 
==3169344== Warning: set address range perms: large range [0x118f4000, 0x22ec3000) (defined)
vex amd64->IR: unhandled instruction bytes: 0x62 0xF1 0xFE 0x8 0x6F 0x67 0x7 0x62 0xF1 0xFE
vex amd64->IR:   REX=0 REX.W=0 REX.R=0 REX.X=0 REX.B=0
vex amd64->IR:   VEX=0 VEX.L=0 VEX.nVVVV=0x0 ESC=NONE
vex amd64->IR:   PFX.66=0 PFX.F2=0 PFX.F3=0
==3169344== valgrind: Unrecognised instruction at address 0x48682c.
==3169344==    at 0x48682C: CrystalLattice (CrystalLattice.cpp:36)
==3169344==    by 0x48682C: qmcplusplus::CrystalLattice<double, 3u>::CrystalLattice() (CrystalLattice.cpp:30)
==3169344==    by 0x47D3D1: qmcplusplus::ParticleSet::ParticleSet(qmcplusplus::DynamicCoordinateKind) (ParticleSet.cpp:77)
==3169344==    by 0x42D610: QMCGaussianParserBase::QMCGaussianParserBase(int, char**) (QMCGaussianParserBase.cpp:165)
==3169344==    by 0x46B99D: DiracParser::DiracParser(int, char**) (DiracParser.cpp:54)
==3169344==    by 0x412C58: main (convert4qmc.cpp:103)
==3169344== Your program just tried to execute an instruction that Valgrind
==3169344== did not recognise.  There are two possible reasons for this.
==3169344== 1. Your program has a bug and erroneously jumped to a non-code
==3169344==    location.  If you are running Memcheck and you just saw a
==3169344==    warning about a bad jump, it's probably your program's fault.
==3169344== 2. The instruction is legitimate but Valgrind doesn't handle it,
==3169344==    i.e. it's Valgrind's fault.  If you think this is the case or
==3169344==    you are not sure, please let us know and we'll try to fix it.
==3169344== Either way, Valgrind will now raise a SIGILL signal which will
==3169344== probably kill your program.
==3169344== 
==3169344== Process terminating with default action of signal 4 (SIGILL): dumping core
==3169344==  Illegal opcode at address 0x48682C
==3169344==    at 0x48682C: CrystalLattice (CrystalLattice.cpp:36)
==3169344==    by 0x48682C: qmcplusplus::CrystalLattice<double, 3u>::CrystalLattice() (CrystalLattice.cpp:30)
==3169344==    by 0x47D3D1: qmcplusplus::ParticleSet::ParticleSet(qmcplusplus::DynamicCoordinateKind) (ParticleSet.cpp:77)
==3169344==    by 0x42D610: QMCGaussianParserBase::QMCGaussianParserBase(int, char**) (QMCGaussianParserBase.cpp:165)
==3169344==    by 0x46B99D: DiracParser::DiracParser(int, char**) (DiracParser.cpp:54)
==3169344==    by 0x412C58: main (convert4qmc.cpp:103)
==3169344== 
==3169344== HEAP SUMMARY:
==3169344==     in use at exit: 1,802,866 bytes in 19,385 blocks
==3169344==   total heap usage: 19,495 allocs, 110 frees, 1,943,417 bytes allocated
==3169344== 
==3169344== LEAK SUMMARY:
==3169344==    definitely lost: 0 bytes in 0 blocks
==3169344==    indirectly lost: 0 bytes in 0 blocks
==3169344==      possibly lost: 0 bytes in 0 blocks
==3169344==    still reachable: 1,802,866 bytes in 19,385 blocks
==3169344==                       of which reachable via heuristic:
==3169344==                         stdstring          : 28,623 bytes in 619 blocks
==3169344==         suppressed: 0 bytes in 0 blocks
==3169344== Rerun with --leak-check=full to see details of leaked memory
==3169344== 
==3169344== For lists of detected and suppressed errors, rerun with: -s
==3169344== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 4 from 4)
Illegal instruction (core dumped)
```

**Comment by camelto2** [comment]
following the errors back from the 
```
==3169344==  Illegal opcode at address 0x48682C
==3169344==    at 0x48682C: CrystalLattice (CrystalLattice.cpp:36)
==3169344==    by 0x48682C: qmcplusplus::CrystalLattice<double, 3u>::CrystalLattice() (CrystalLattice.cpp:30)
==3169344==    by 0x47D3D1: qmcplusplus::ParticleSet::ParticleSet(qmcplusplus::DynamicCoordinateKind) (ParticleSet.cpp:77)
==3169344==    by 0x42D610: QMCGaussianParserBase::QMCGaussianParserBase(int, char**) (QMCGaussianParserBase.cpp:165)
==3169344==    by 0x46B99D: DiracParser::DiracParser(int, char**) (DiracParser.cpp:54)
==3169344==    by 0x412C58: main (convert4qmc.cpp:103)
```

it is getting to the constructor for DiracParser, which invokes the QMCGaussianParserBase constructor. QMCGaussianParserBase has a member ParticleSet IonSystem, which invokes its constructor and so on until we get to the CrystalLattice constructor. It is failing at the line 

G = R; in the CrystalLattice constructor. 

The gamess converter for example should follow the same code path once it gets into the QMCGaussianParserBase constructor, so I'm not sure I understand why the dirac option is resulting in this "illegal opcode" failure while others aren't.

**Comment by ye-luo** [comment]
@camelto2 I just turned on [address sanitizer](https://qmcpack.readthedocs.io/en/develop/external_tools.html?highlight=address%20sanitizer#sanitizer-libraries)
```
$ ctest -R converter_test_Bi_dirac --output-on-failure
Test project /home/yeluo/opt/cleanup/qmcpack/build_gnu_real_asan
    Start 52: converter_test_Bi_dirac
1/1 Test #52: converter_test_Bi_dirac ..........***Failed    0.16 sec
Return code nonzero:  1
Stderr not emptry 
=================================================================
==3759307==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffe68d1f2fc at pc 0x7fc5eae17a6d bp 0x7ffe68d1d780 sp 0x7ffe68d1cf28
READ of size 15 at 0x7ffe68d1f2fc thread T0
    #0 0x7fc5eae17a6c  (/usr/lib/x86_64-linux-gnu/libasan.so.5+0x67a6c)
    #1 0x5556f89e52e2 in parsewords(char const*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /home/yeluo/opt/cleanup/qmcpack/src/Utilities/SimpleParser.cpp:69
    #2 0x5556f87cb220 in DiracParser::getSpinors(std::istream&) /home/yeluo/opt/cleanup/qmcpack/src/QMCTools/DiracParser.cpp:509
    #3 0x5556f87e95fb in DiracParser::parse(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /home/yeluo/opt/cleanup/qmcpack/src/QMCTools/DiracParser.cpp:204
    #4 0x5556f86a4af7 in main /home/yeluo/opt/cleanup/qmcpack/src/QMCTools/convert4qmc.cpp:288
    #5 0x7fc5e95e30b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #6 0x5556f86a924d in _start (/home/yeluo/opt/cleanup/qmcpack/build_gnu_real_asan/bin/convert4qmc+0x6324d)

Address 0x7ffe68d1f2fc is located in stack of thread T0 at offset 5484 in frame
    #0 0x5556f87c7a7f in DiracParser::getSpinors(std::istream&) /home/yeluo/opt/cleanup/qmcpack/src/QMCTools/DiracParser.cpp:416
```

**Comment by camelto2** [comment]
@ye-luo I was able to reproduce the asan error you were seeing, and it had to do with my usage of sscanf. I decided to change the way I parse this using the parsewords function and a check to make sure it actually finds the right number of strings in the line. This approach fixes things on my end, so hopefully everything pass this time

**Review by ye-luo — APPROVED** [review]

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**